### ISSUE TITLE
duplicate challenge tasks in participant's task list

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#11295]
[EDIT: See the first comment below for an important update.]

@SuperSaraA (dbf5390d-6a57-42ad-ae14-af8fd8e570ca) has reported a recurring problem with duplicate challenge tasks appearing on the website. In the most recent case at least the duplicates later disappeared by themselves (unless @sabrecat cleaned them up and didn't post about it?). All the details we have so far are below.

For the moment, I'm marking this as needs investigation. If it happens again and I can capture details of the tasks before they disappear, I'll update this issue.

@SuperSaraA: "a continuation of a problem that I have had more than once. I am somehow in one challenge more than one time. I have no idea how this keeps happening to me. Lucky for me there are not any daily tasks to this challenge. Is there any way that you could please get rid of all the extra copies of the to-do tasks from the Celebrate August challenge (I think I counted that there are about 7 or 8 extra copies of all the tasks)? If this is something you can't do, I guess I can just leave the challenge and re-sign up."

@shanaqui: "have we ever logged this as a bug? I'll tag in @Sabrecat to take a look at cleaning this up for you"

@SuperSaraA [roughly a day after the initial report]: "The extra tasks are now gone. I wonder if it happens when I am trying to join a challenge with a not so great internet connection."

@Alys: "Were those extra tasks visible on the Android app or website or both?"

@SuperSaraA: "I could see all the extra to dos on the website but I'm not sure about on the android app. I remember thinking my to do list was getting way too long because the little scroll line on the right side was so tiny (and I was starting to forget the stuff at the top by the time I would get to the bottom of the list) . So I want onto the website version to look over my tasks in more detail and that is when I for sure saw all the extra copies of the tasks. Sorry, I can't remember more! If it happens again, I will be sure to take notes of if I can see it on which platform."

**Comment by Alys** [comment]
An update from SuperSaraA:
"Hi again. I just looked at Habitica on the website and all the extra challenge to do copies are still there. The challenge in question is Celebrate August! [here](https://habitica.com/challenges/c21661c0-e0f2-4cef-8b10-bd8b31b4270e). I have just checked that these extra copies are not on my android app. They are not. I can only see one copy of the to dos on my phone version."

I've confirmed that multiple copies of that challenge's tasks exist. For example, here's multiple copies of just one of the challenge's To-Dos:
```
{"_id":"a22b5ca7-e0a4-45fd-a473-b025eab06cc0","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
{"_id":"ce1783f2-6b4e-4111-abc7-2f371dfe6fe5","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
{"_id":"3f434cd5-f775-437e-b6cf-5a448db3cc58","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
{"_id":"02af304b-5d31-4590-a624-31e468a52037","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
{"_id":"918cd346-490b-4837-842f-9aaa6d4602f4","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
{"_id":"1cf0af87-8bee-42a7-a277-27ededdc18c6","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
{"_id":"e77b56a6-ffae-4ce6-ba7e-10bfe5a30489","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
{"_id":"b03c0d5e-8424-4529-9abd-bdedf6bc6022","challenge":{"taskId":"5baf64cc-de69-4019-a4fd-9911eb1d35fd","id":"c21661c0-e0f2-4cef-8b10-bd8b31b4270e"},"type":"todo","tags":["c21661c0-e0f2-4cef-8b10-bd8b31b4270e"],"text":"August 3rd: National Watermelon Day ","userId":"dbf5390d-6a57-42ad-ae14-af8fd8e570ca"}
```
Note that this can't happen by leaving the challenge and rejoining it. If you leave and choose the option to keep the tasks, those tasks become unlinked from the challenge, and they stay unlinked if you join again (I've just tested to confirm that). So those duplicate tasks above must have all been created by Habitica when joining a challenge. 

I'm marking this as help wanted.

**Comment by Alys** [comment]
There's another case of this reported by @jjgame83 (d13df193-6267-4e87-b989-f2a5e6d3f264):

_"I've noticed that I have three of the same challenge daily on website titled "Good Work! I noticed myself doing something well and gave myself a compliment at least once today" [challenge ID 5977368d-6814-4f0e-ac0e-e17684fae886]. On my app there is only one that I click, but on the website, only one of the daily shows 10 day streak while the two 'shadow' dailies show 0. I'm also suspecting that I might not be getting my perfect day achievement because of this, but I don't have evidence to support this. I will be keep tracking from today, but it would be great if someone could verify from the back end. Thanks
Edit: I've checked that the same issue happens for the habit and to-do tasks under this same challenge."_

I've confirmed that each challenge task exists three times in his account, all created within the same second so this must have happened from just one attempt to join the challenge.

Only one copy of each task appears in his tasksOrder array, which explains why he wasn't able to see the duplicates on the app (due to this Android app bug: https://github.com/HabitRPG/habitica-android/issues/603). I assume the same was true for SuperSaraA; I hadn't made the tasksOrder connection when I was looking at her case.

**Comment by Alys** [comment]
This happened to SuperSaraA again, after joining the challenge using the Android app:

"I am having trouble with duplicated items from a challenge again. I joined the Habitica official back to school challenge a few days ago (using my android app) and today see on the website version that I have two copies of all the tasks. I do only see one set on my android app. Lucky me there are no daily tasks, but I thought you would like to know."

Again the duplicate tasks were not in the tasksOrder array, meaning that there would have been issues with sorting tasks (and also there would have been invisible Dailies on the app if the challenge had contained Dailies).

**Comment by Alys** [comment]
As a note for admins: The fact that the duplicates don't appear in tasksOrder makes it pretty easy to identify and remove the duplicates, although you do need to check that other, unrelated tasks haven't been affected by the tasksOrder bug.

**Comment by Alys** [comment]
SuperSaraA has reported this again. For a challenge that contained 31 tasks, 123 duplicate tasks were added (in addition to the 31 that should be there).

I'm raising this to a high priority because duplicate tasks can make your task list unusable, especially when there's that many extra and the tasksOrder aspect prevents you reordering your tasks.

This time I've been able to find some information that might help to explain the bug.
Loggly has three instances of the error below all within the same second while SuperSaraA was joining the challenge:

```
originalUrl: /api/v4/challenges/703b2ea6-42a2-4b79-ae76-f6c871044aa7/join  
   [that's the ID of the challenge that exhibited the problem]

message:VersionError: No matching document found for id "dbf5390d-6a57-42ad-ae14-af8fd8e570ca"
  version 437 modifiedPaths "auth, auth.timestamps, auth.timestamps.updated, challenges, tags, tasksOrder, tasksOrder.todos"
	...
	at async handler (/usr/src/habitrpg/website/transpiled-babel/controllers/api-v3/challenges.js:266:19)
```

In addition, examining the challenge's data shows that in the challenge's own tasksOrder object there are two duplicate IDs (i.e., the same challenge task appears there twice). That data is shown below but I'm about to fix it in the database so it will no longer be visible there. I've logged a separate issue to get that fixed: https://github.com/HabitRPG/habitica/issues/11385

```
{
    "_id": "703b2ea6-42a2-4b79-ae76-f6c871044aa7",
    "tasksOrder": {
        "habits": [],
        "dailys": [],
        "todos": [
            [12 non-duplicated task IDs removed for brevity],
            "bd3f6dfe-ef63-44dd-ad46-79739bf06e8b",
            "1c1723d5-4e76-4c29-9044-31012130c99f",
            "1c1723d5-4e76-4c29-9044-31012130c99f", [duplicate]
            "9540175b-9cfc-4775-a366-8b0cc8bb0492",
            "fe972675-958b-4ee7-a209-f99ca779277c",
            "d62d4ada-9974-4d1e-a749-5a418045bbe6",
            "2d9e89d6-a7ca-4fce-b300-bbbee313f238",
            "2d9e89d6-a7ca-4fce-b300-bbbee313f238", [duplicate]
            "d69a8ebb-63ae-4186-811a-a6ed0b158de7",
            [12 non-duplicated task IDs removed for brevity],
        ],
        "rewards": []
    },
}
```

I'll be examining at least some of the challenge's other participants (327 of them unfortunately) to see if they also are affected by the duplicate tasks bug.

**Comment by cvuorinen** [comment]
I looked into this a bit and I was able to reproduce it by tapping the "Join" button quickly multiple times on the Android app. So seems like a race condition when multiple identical requests come in at the same time and start processing the request. They go through the check if user is already a participant in the challenge before the first request has gotten to the point where it saves the user data, and then they all start the process of adding tasks etc.

One additional thing I noticed is that the challenges `memberCount` also gets out of sync, since it's just an integer that is incremented each time the join is processed.

<details>
 <summary>In case any admins want to investigate, click here for details</summary>

> My User ID: `fa756acf-9127-4a3c-8dac-8ff627d30073`
> Challenge ID: `d1c87633-0a7d-4ca0-bd9f-80fcdfdecb29`
> There are currently actually 4 participants, but the count says 7.
> I have the challenge habits and todos twice in my task list (and I can see them all in my Android app).
> I did join this same challenge two times, because on my first try I only got the tasks once. Then I left the challenge and removed the tasks and tried again.
</details>

I also looked into joining a challenge through the website, but the problem is not present there, as the Join button gets hidden as soon as you click it once.

So the easiest way to fix this would probably be to prevent the Android app from sending multiple concurrent requests, by disabling the button while a request is already in progress. The problem with this approach is that it requires users to update the app to get the fix.

I think there are also ways to fix this in the server side. I looked into making atomic updates with MongoDB and it should be possible to combine the check of user already added to the challenge with the actual adding operation. Doing it this way would require the adding operation to be done first, after all the safety checks, and before starting to add the tasks etc. and to also check that the `modifiedCount` return value of Mongo's update operation is 1 (and throw error if it's 0 so that it will stop executing, as another request already added it).

**Comment by cvuorinen** [comment]
Also found this old PR https://github.com/HabitRPG/habitica/pull/7787, where the same issue was fixed on the website. There's more discussion related to atomic updates on MongoDB and Mongoose, but the "solution" in that PR is to just let simultaneous request go through and let the last one overwrite (which is fine for just adding the challenge to users `challenges` array, but does not solve the issue with tasks etc.)

**Comment by Alys** [comment]
@cvuorinen Awesome, thanks so much for this investigation! 

As a result of it, your account had a mild case of the tasksOrder sorting bug but I've fixed that for you. If you'd like to remain in the challenge but keep only one set of the challenge tasks, post again here and I'll remove the extra set for you. If you don't want any of them, leaving the challenge and choosing the option to remove the tasks should get rid of all of them (but post here if it doesn't).

I've also fixed the memberCount in the challenge - thanks for advising us about that aspect of the bug.

My personal feeling is that server-side would be the best place to fix this issue, even though that would take more work, since any client-side fixes would have to be done in all clients (potentially including third-party clients) and would be depended on the fixes remaining in the code in future (that's not necessarily reliable since the clients undergo frequent changes).

Are you interested in working on a server-side fix? If so, is your suggestion (copied below) the fix you'd want to do? If you'd like to implement that yourself, post here to say so but you might like to wait two or three days to see if the staff comment on it in case they'd prefer a different approach. Since it's the weekend now it might be a day or two before they see these updates.

> I looked into making atomic updates with MongoDB and it should be possible to combine the check of user already added to the challenge with the actual adding operation. Doing it this way would require the adding operation to be done first, after all the safety checks, and before starting to add the tasks etc. and to also check that the modifiedCount return value of Mongo's update operation is 1 (and throw error if it's 0 so that it will stop executing, as another request already added it).

**Comment by paglias** [comment]
Just a reminder that we're using Mongo 3.6 not 4.x so we can't use transactions to update multiple docs atomically

**Comment by cvuorinen** [comment]
@Alys yes, I can work on the fix.

@paglias no need for transactions or multi doc atomic operations, at least I don't think so. My idea is to just update the `user.challenges` array atomically, and then check if should continue with the other operations or not.

**Comment by paglias** [comment]
@cvuorinen I'm happy to see any fix for this until we're able to use transactions in Mongo 4 but I want to point out that a similar fix was attempted previous and updating the `user.challenges` array atomically won't solve the issue unless the query is made with the condition that if the challenge id is already present in the `user.challenges` array. 

Otherwise race conditions can still happen, for example if you send at the same time two updates like this:

```
user.update({$push: {challenges: challengeId}})
```

the update is atomic but nothing prevent it from happening twice. There's `$addToSet` that can help here https://docs.mongodb.com/manual/reference/operator/update/addToSet/

See also this old PR https://github.com/HabitRPG/habitica/pull/7787

**Comment by paglias** [comment]
In the meantime I'll increase the priority on moving to mongo 4

**Comment by cvuorinen** [comment]
@paglias yeah I read through that PR already (I linked to it yesterday from here).

I was thinking to update the challenges array with a query that also checks if the id already exists in the array, so that it will only add the value once, and then check the modified count that Mongo returns to see if it modified anything or not. That way this operation is atomic, and by checking the return value we can decide to continue or not. 

So, to recap:

1. Move adding challenge id to `user.challenges` right after the sanity checks, before starting to do other db operations.
2. Update atomically with a condition to check if already exists (see query below).
3. Check modified count, if query did not modify (modifiedCount=0), then exit here (another request already added the challenge).
4. Continue with rest of the operations, like adding challenge tasks & tags.

And the update query would be something along the lines of:
```
db.users.update(
    { 
        _id: "user-id", 
        challenges: { $nin: [ "challenge-id" ] } 
    }, 
    { $push: { challenges: "challenge-id } } 
)
```

If the challenges array already contains "challenge-id", then the where condition will not match and the value will not get pushed twice, and Mongo returns modified count 0. If the where condition matches the user, then the "challenge-id" is not in the array and it gets pushed and Mongo returns modified count 1.

**Comment by paglias** [comment]
@cvuorinen I'm not against this solution until we can use transactions but I want to point out that, yes this way it prevents the code from being called twice but it doesn't stop the `user.update` to succeed while the other operations like adding challenge tasks to the user to fail later resulting in incomplete data.

That being said if you want to open a PR making this change we'll approve it

**Comment by cvuorinen** [comment]
@paglias I'm not sure I follow what you mean by this:
> but it doesn't stop the `user.update` to succeed while the other operations like adding challenge tasks to the user to fail later resulting in incomplete data.

Or do you mean some other possible error condition that might happen later in the process (not related to this issue of multiple concurrent requests)? If yes, then I agree transaction is the way to solve it.

**Comment by paglias** [comment]
> Or do you mean some other possible error condition that might happen later in the process (not related to this issue of multiple concurrent requests)? If yes, then I agree transaction is the way to solve it.

This.

Let's see what @Alys think is best to do while we wait for transactions

**Comment by Alys** [comment]
I like @cvuorinen's approach and think it's worth doing.
This bug that gives multiple tasks and tasksOrder errors is more troublesome than failing to have tasks added. Anyone suffering the latter bug can fix it by leaving and rejoining the challenge.

**Comment by paglias** [comment]
In this case @cvuorinen a PR would be great, thanks! Let me know if you need any help or have any question regarding that
